### ISSUE TITLE
ec2_vpc_routetable produces "expected params[:internet_gateway_ids][0] to be a string" error

### ISSUE THREAD

**Issue by lynxdigital** [puppetlabs/puppetlabs-aws#192]
Whenever I use the ec2_vpc_routetable function I receive the error:

```
Debug: Prefetching v2 resources for ec2_vpc_routetable
Debug: Storing state
Debug: Stored state in 0.01 seconds
Debug: Using settings: adding file resource 'rrddir': 'File[/var/lib/puppet/rrd]{:path=>"/var/lib/puppet/rrd", :mode=>"750", :owner=>"puppet", :group=>"puppet", :ensure=>:directory, :loglevel=>:debug, :links=>:follow, :backup=>false}'
Debug: Finishing transaction 29631360
Debug: Received report to process from xxx.compute.internal
Debug: Processing report from xxx.compute.internal with processor Puppet::Reports::Store
Error: Could not run: Puppet detected a problem with the information returned from AWS
when looking up ec2_vpc_routetable in ap-southeast-2. The specific error was:

expected params[:internet_gateway_ids][0] to be a string

Rather than report on ec2_vpc_routetable resources in an inconsistent state we have exited.
This could be because some other process is modifying AWS at the same time.
```

Have tried a number of different class names,etc. and always end up receiving this error. Any idea what might be causing it and how to fix it.

**Comment by garethr** [comment]
Do you see this when running `puppet resource` or when running against a manifest? If the latter would you be able to post a gist of your manifest?

This might be the module being too heavy on the validation for the specific param.

**Comment by rfletcher** [comment]
I had the same problem. It's triggered when a route table includes a route to a NAT instance. Here's a partial response from `aws ec2 describe-route-tables` showing a NAT route, which lacks the expected `GatewayId` property:

````json
...
  {
      "GatewayId": "local", 
      "DestinationCidrBlock": "10.0.0.0/8", 
      "State": "active", 
      "Origin": "CreateRouteTable"
  }, 
  {
      "Origin": "CreateRoute", 
      "DestinationCidrBlock": "0.0.0.0/0", 
      "InstanceId": "i-abcd1234", 
      "NetworkInterfaceId": "eni-abcd1234", 
      "State": "active", 
      "InstanceOwnerId": "954578106580"
  }
...
````

Here's my quick and dirty patch to avoid the exception: https://github.com/rfletcher/puppetlabs-aws/commit/0b96095a5c8e1756cd07a20df129a899a7c30149

**Comment by rfletcher** [comment]
I had the same problem. It's triggered when a route table includes a route to a NAT instance. Here's a partial response from `aws ec2 describe-route-tables` showing a NAT route, which lacks the expected `GatewayId` property:

````json
...
  {
      "GatewayId": "local", 
      "DestinationCidrBlock": "10.0.0.0/8", 
      "State": "active", 
      "Origin": "CreateRouteTable"
  }, 
  {
      "Origin": "CreateRoute", 
      "DestinationCidrBlock": "0.0.0.0/0", 
      "InstanceId": "i-abcd1234", 
      "NetworkInterfaceId": "eni-abcd1234", 
      "State": "active", 
      "InstanceOwnerId": "01234567890"
  }
...
````

Here's my quick and dirty patch to avoid the exception: https://github.com/rfletcher/puppetlabs-aws/commit/0b96095a5c8e1756cd07a20df129a899a7c30149

**Comment by lynxdigital** [comment]
rfletcher - Thanks. Your patch did the trick and got rid of the error. 

I've hit another problem, but I'll create a new issue for it if I can't find any other issues that have been raised against it.

**Comment by garethr** [comment]
I'll keep this open until we get chance to fix in the main codebase. Thanks for the discussion and report, much appreciated.

**Comment by earsdown** [comment]
+1

**Comment by farkasmate** [comment]
+1

**Comment by daveseff** [comment]
+1

**Comment by daveseff** [comment]
The patch from @rfletcher works for me. @rfletcher are you going to do a PR or should I? I don't want to take credit for this fix but would like to see it in the upstream soon.

**Comment by rfletcher** [comment]
@daveseff Did you try the fix in PR #190? That's already been merged. (The error message was the same there, but I don't recall if that change resolved this issue.)

**Comment by daveseff** [comment]
Thanks. I was working on a system that has been up for a while so it may not have had that PR fix yet. I think we're all good now.

**Comment by ahenroid** [comment]
@lynxdigital @rfletcher @daveseff I'm doing some house cleaning. I believe this was fixed by #190. Can I close this issue?
